### PR TITLE
APEXCORE-775  do not update the conf object in evalDFSRootDir when called for APEX_APP_DFS_ROOT_DIR

### PR DESCRIPTION
**Pull request by sanjaypujare**
@PramodSSImmaneni  pls review and merge as appropriate

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java]: Should a similar thing be done with APEX_DFS_ROOT_DIR, how is this property that is set on the conf used after the call, given that the method is also returning this path

**Comment by pramodin** [comment]
Take a look at the newFileSystemInstance method as well, I believe it should be calling getApexDFSRootDir

File: engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java
Comment (sanjaypujare) [reply engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java]: @PramodSSImmaneni  As mentioned elsewhere, we cannot do the same thing for APEX_DFS_ROOT_DIR since the value is per Hadoop login user so cannot cache it in the conf object

**Comment by sanjaypujare** [comment]
@PramodSSImmaneni Regarding newFileSystemInstance we may need another function since the function is called in 2 different contexts: in one, it needs to return what getDTDFSRootDir() computes (basically a fixed directory used for config etc) and in the other it needs to return what getApexDFSRootDir() computes (an application run time directory based on current Hadoop user). Just changing newFileSystemInstance() can break a number of things?

Analyzing this further, the calls to newFileSystemInstance() in `com.datatorrent.stram.client.StramClientUtils.addDTSiteResources(Configuration)` and `com.datatorrent.stram.cli.ApexCli.init()` need to use the current logic (i.e. `getDTDFSRootDir`) but all the other calls in Apex-core need to use the value returned by `getApexDFSRootDir()`. There are also probably cases of other3rd party software calling `StramClientUtils.newFileSystemInstance()` which relies on the current logic so it won't be prudent to change the behavior now.

So I suggest creating a new function to support this new behavior and use this function wherever required.

**Comment by sanjaypujare** [comment]
@PramodSSImmaneni  any comments on my comment? Should I go ahead make the changes for newFileSystemInstance based on my suggestion

File: engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java
Comment (pramodin) [reply engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java]: Can the property be passed as a parameter and can this method capture more of the common processing between getApex and getDT methods such as reading the property value and getting the default root folder if the value is empty. The username and home folder can be passed from the respective get.. methods.

**Comment by pramodin** [comment]
@sanjaypujare I agree with your comments on newFileSystemInstance, but what was not clear in my statement was that I was more concerned with the method repeating what get RootDir methods were already doing.

File: engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java
Comment (sanjaypujare) [reply engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java]: Done. Pls check

**Comment by sanjaypujare** [comment]
@PramodSSImmaneni  I was able to refactor and reuse a method from both evalRootDir and newFileSystemInstance methods. Pls check. 

If okay, could you merge?

**Review by vrozov — CHANGES_REQUESTED** [review]
Tests must be fixed before PR is merged.

**Comment by sanjaypujare** [comment]
Tests seem unrelated to this PR change. Tests pass in my workarea.

**Comment by sanjaypujare** [comment]
Jenkins test please

**Comment by sanjaypujare** [comment]
Closing to retest

**Comment by sanjaypujare** [comment]
reopening to retest

File: engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java]: Remove the method name from the log message (already provided by logger).

Comment (sanjaypujare) [reply engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java]: fixed ( in both cases)

**Comment by sanjaypujare** [comment]
@vrozov  I have checked multiple times and all tests pass in my workarea so the failures on Jenkins and Travis seem unrelated to these changes (and they are different between the 2). You can check out this branch and run the tests on your machine to verify.

**Comment by sanjaypujare** [comment]
The travis build passed - and the Jenkins build was aborted.

**Review by vrozov — CHANGES_REQUESTED** [review]
Jenkins build continues to fail and it fails in the same set of test. See https://builds.apache.org/job/Apex_Core_PR/1063/ and https://builds.apache.org/job/Apex_Core_PR/1067/

**Comment by sanjaypujare** [comment]
Looking at https://builds.apache.org/job/Apex_Core_PR/1064 and https://builds.apache.org/job/Apex_Core_PR/1065 those same tests pass (https://builds.apache.org/job/Apex_Core_PR/1064/org.apache.apex$apex-engine/testReport/com.datatorrent.stram/StramMiniClusterTest/  and https://builds.apache.org/job/Apex_Core_PR/1065/org.apache.apex$apex-engine/testReport/com.datatorrent.stram/StramMiniClusterTest/) and in 1064 a completely different set of tests fail :

com.datatorrent.stram.engine.GenericNodeTest.testDoubleCheckpointExactlyOnce
com.datatorrent.stram.engine.InputNodeTest.testDoubleCheckpointExactlyOnce

I still believe the tests are unrelated to the PR and are failing due to instability.

**Comment by sanjaypujare** [comment]
retest please

**Review by vrozov — CHANGES_REQUESTED** [review]
Another Jenkins build https://builds.apache.org/job/Apex_Core_PR/1068/ that failed for the same set of tests.

**Comment by sanjaypujare** [comment]
If you think the tests are failing due to code change in this PR, why are they passing on Travis and also on Jenkins in 1064 and 1065? Also they are passing on my local machine as you too can verify on your local machine. IS there way we can run Jenkins tests on the master branch to compare?

**Comment by vrozov** [comment]
The question to ask is not why the tests pass locally or on travis-ci, the question to ask is why those tests failed on Jenkins and as long as the same test(s) failed more than once, I think that the failure is related to PR changes. It is the contributor responsibility to prove otherwise.

**Comment by sanjaypujare** [comment]
testing again

**Review by vrozov — CHANGES_REQUESTED** [review]
Just to clarify - it does not matter whether the next Jenkins build will succeed or fail, at this point without RCA for the prior Jenkins build failures I am -1 on the PR.

**Comment by sanjaypujare** [comment]
Okay I am going to open a different PR with no code change to see if that Jenkins build has these issues or not - since there is no other way to debug these things (I cannot even reproduce the test failures)

**Comment by sanjaypujare** [comment]
@vrozov Pls take a look at https://builds.apache.org/job/Apex_Core_PR/1070/#showFailuresLink . This was triggered by https://github.com/apache/apex-core/pull/571 . Let me know what the root cause is for those failures and I can report the root cause for the failures in this PR.

**Comment by sanjaypujare** [comment]
Just for the record: the PR #571 has the following change:

```
 engine/src/main/java/com/datatorrent/stram/StramClient.java
 @@ -93,6 +93,7 @@
   * <br>
   *
   * @since 0.3.2
 + *
   */
  @InterfaceAudience.Public
  @InterfaceStability.Unstable
```

And the failures reported for that PR are:
```
  | No changes.
-- | --
  | GitHub pull request #571 of commit 8930650f6d5ff77542d14f1d1a718759dd802638, no merge conflicts.
  | This run spent:7.5 sec waiting in the queue;11 min building on an executor;12 min total from scheduled to completion.
  | Test Result (2 failures / ±0)
com.datatorrent.stram.StramMiniClusterTest.testSetupShutdown
com.datatorrent.stram.StramMiniClusterTest.testAddAttributeToArgs
```

**Comment by sanjaypujare** [comment]
@vrozov  given the information, pls give me a technical reason for your -1 for this PR or it will be considered void

**Review by vrozov — CHANGES_REQUESTED** [review]
I am -1 on this PR and on any other PR where build continues to fail. Prior to merging any new PR, build issues must be fixed.

**Comment by sanjaypujare** [comment]
@vrozov hmmm, your own PR #570 has a build failure. Is there a resolution for it? 

#568 also has a build failure but I don't see any comments from you about -1 because of build failure. Why single this one out? Also since all of these PRs have been before this PR, shouldn't those authors be fixing the build failures?

**Comment by tweise** [comment]
This highlights the benefits of working CI, that very few are willing to put effort in. Status quo not helping.

**Comment by sanjaypujare** [comment]
@tweise thanks for your comments. You are right that we should be willing to put the effort to have a working CI and weed out unstable tests. But that should be done as a separate PR and JIRA. Singling this PR out for the build failures when that was not an issue for previous PRs is not justified. What do you think about -1 knowing all the facts.

**Comment by tweise** [comment]
I have been in the situation to fix CI issues caused by other contributors that had nothing to do with my PR. And I also know that when the same contributors say "later" it usually means never.

**Comment by sanjaypujare** [comment]
@tweise there are 2 different things: new CI issues caused by other contributors that impact you versus persistent unstable test failures which is the case here. We need to find a way to encourage contributors to help with CI maintenance tasks. Unjustified -1 on PRs is not the way to do it.

Anyway since the discussion has started, should we -1 all the PRs in Apex-core until the unstable tests are fixed?

**Comment by tweise** [comment]
In other communities I see that contributors take up CI issues as high priority because they know that it affects everyone and the working CI benefits the project as a whole.

**Comment by sanjaypujare** [comment]
@tweise that is a general answer to a specific question.

Something to think about: we need to encourage contribution to Apex instead of discouraging it. Hiring someone (paying them) to fix CI issues is easy but difficult in a voluntary community such as this. But I can't be getting -1 on my PRs for that problem.

**Comment by vrozov** [comment]
@sanjaypujare regarding #570 and #568 - the same rule applies and I am looking into why the build failed and do not plan to pull the trigger on #568 either. The difference between #570 and your PR is that you asked PR to be merged completely ignoring that both Travis and Jenkins build failed.

**Comment by sanjaypujare** [comment]
@vrozov  granted to some extent, but isn't that the most common process? I add a comment "asking to merge the PR" with my latest commit (squashed etc) at which point the CI build is triggered and the results are known 10-15 minutes later. BTW, I submit my commit ONLY after I have verified all tests pass on my local machine.

**Comment by vrozov** [comment]
@sanjaypujare There was sufficient time to look into test failure between your last commit/comment and my request to fix unit tests before PR is merged. The fact that the build passes on your local machine does not mean that it would pass in Travis or Jenkins.

**Comment by sanjaypujare** [comment]
@vrozov we can endlessly discuss the minutiae of PR/CI process which I have no interest in doing. Looking at the big picture I would like the review process to be objective and technical and on the whole it should encourage contribution instead of actively discourage it. There are many open PRs with CI failures but no comment from you or anybody else regarding that.

**Comment by sanjaypujare** [comment]
This statement makes little sense: "The fact that the build passes on your local machine does not mean that it would pass in Travis or Jenkins."

So I cannot take anything for granted and stop all software development activity because a program running on my machine is no guarantee that it will run the same way on a user's machine.

**Comment by tweise** [comment]
> Something to think about: we need to encourage contribution to Apex instead of discouraging it. Hiring someone (paying them) to fix CI issues is easy but difficult in a voluntary community such as this.

It's not done by hiring someone, but as part of the contribution process, collectively by the contributors. That's not much different than in companies that encourage sound processes and ownership, where things are not thrown over the fence but owned by engineers that take pride in their work and understand the value things like CI provide.

**Comment by sanjaypujare** [comment]
@tweise can't disagree with your thoughts. However see it from my point of view just for this PR. There are users, customers etc who are waiting for this functionality and I take great pride in the work I have done here. My focus is on getting this done in a timely manner and at highest quality. Unlike many others I add plenty of unit tests which is what I have done here too when I incorporated @PramodSSImmaneni's comments because I was worried about regressions. After @vrozov's comments I also conscientiously investigated Jenkins failures and ruled out my change as the root cause. I also proved thru another PR that the failures are not caused by my change. But @vrozov doubled down on his -1 instead of acknowledging my proof.

Anyway, if this is really the way things are, I don't see a solution for myself. My employer pays me to do this work and if I am unable to do it, I just have to find something else to do. Somebody else will pick up this JIRA who might have better luck than me with reviews and -1's.

**Comment by tweise** [comment]
@sanjaypujare assuming the PR is merged, what happens after that? How will we not have the same discussion and frustration next time?

**Comment by vrozov** [comment]
> @vrozov we can endlessly discuss the minutiae of PR/CI process which I have no interest in doing. 

There is no need to discuss it, instead, it will be good if you review and follow in the future [the contributing guidelines](http://apex.apache.org/contributing.html) and namely item 5 in "Opening Pull Requests". It clearly mandates "Check the status of the pull request and ensure the Travis CI build is successful. If not, inspect the linked build log for details." Hopefully, this also clarifies for you what I mean by "The fact that the build passes on your local machine does not mean that it would pass in Travis or Jenkins".

> So I cannot take anything for granted and stop all software development activity because a program running on my machine is no guarantee that it will run the same way on a user's machine.

It will be good to review why projects use Jenkins, Travis and other build automation platforms and do not rely on the builds done on a developer machine, it is 101 of software development best practices. 

> There are users, customers etc who are waiting for this functionality and I take great pride in the work I have done here. My focus is on getting this done in a timely manner and at highest quality.

Is not this PR a follow-up PR for #545? Was the #545 (#474, #520 that also required a follow-up PR)  done at the highest quality as well?

**Comment by sanjaypujare** [comment]
@tweise not sure. If you are hinting that somebody needs to look at CI issues eventually and it might as well be me, I don't mind doing it but getting me to do it by raising a -1 is not the way to do it. I am happy to get a -1 for a legitimate bug or issue in the PR.

**Comment by sanjaypujare** [comment]
@vrozov regarding your first 2 points: I did accept your point about build failures and looked at those failures on Jenkins and also ran those tests on my machine to reproduce them and analyze them. I ran the tests in the IDE and put breakpoints to see how my code change could have caused those failures. The bottom line is that those failures were not reproducible on my machine and were not caused by my code.

Regarding your second point about software development 101, what became obvious is that I wouldn't trust test failures on the Jenkins machine and you wouldn't trust test passes on my local machine (or even the Travis-CI machine). If a test is not reliable, whether a test passes or fails has no value. But even after I proved that the test failure is not related to my PR change, you wanted to hold up the PR with a -1 insisting that the test failures need to be fixed. If you want me to fix them there are other ways to do it.

Your last point will be addressed in my next comment.

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (vrozov) [inline engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java:200]: Why is YarnConfiguration necessary? It is cast to Configuration anyway.

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (vrozov) [inline engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: Is not this a global change? Any test must be written in such way that it does not affect other tests.

**Comment by sanjaypujare** [comment]
@vrozov regarding your comment about "highest quality" and the PRs you cited, yes, you are right about the quality lapses as evidenced by the follow-up PRs (except #520) and ultimately I take the responsibility. However it is also true that there aren't enough unit tests to detect regressions. Also something needs to be said about our lengthy review process such as this. This process is neither helping with quick addition of code and features nor is it preventing regressions or bugs if that was one of the goals. I can't even rule out cases where incorporating review comments has in fact introduced regressions esp. if there are no unit tests.

**Comment by vrozov** [comment]
@sanjaypujare I did not see any activity on the PR except for the request "If okay, could you merge?" till I brought attention to the failed tests and even after that, there was no attempt to analyze and fix the problem. All your subsequent comments pointed towards that the most common process is to ignore Travis and Jenkins failures. All that caused me to -1 the PR.

It is not a question whether I trust Travis or Jenkins. It is the community that trust Travis and Jenkins more than your local build as your local environment may have uncommitted changes, unknown java version, and the build is not repeatable by anybody else except you.

I can't help with the length of the PR review especially when the overall quality of the submitted PRs is far from what I would expect.

IMO, it is a sloppy route to blame somebody for missing unit tests or comments for submitted PRs that introduced regressions. If they are missing, add them. The same applies to Jenkins or Travis. If they fail, fix the test or script. If review comment is wrong, it is the contributor responsibility to point the reviewer why it is wrong.

At this point I believe that subject why I -1 the PR is beaten to death and I don't plan to reply on it any longer (ok to discuss code changes). For this or any other PR to be merged, Jenkins build must be fixed.

**Comment by tweise** [comment]
> If you are hinting that somebody needs to look at CI issues eventually and it might as well be me, I don't mind doing it but getting me to do it by raising a -1 is not the way to do it. I am happy to get a -1 for a legitimate bug or issue in the PR.

That someone needs to look at the CI issues is clear. How about those now affected by past negligence collaborate to resolve the issues that block their work?

**Comment by sanjaypujare** [comment]
@tweise yes, I am happy to collaborate. We can continue the discussion elsewhere (dev list) since it is not related to the PR

**Comment by sanjaypujare** [comment]
@vrozov As mentioned in my previous comment I am happy to collaborate on fixing CI issues wherever they occur. If they occur in the context of my PR's changes, I will address them as part of the PR. But if they are unrelated or generic in nature, I don't want to waste my time trying to find a non-existent link between my code and the CI issues, as it happened in this case. I hope you don't find that expectation unreasonable.

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (sanjaypujare) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java:200]: Good point. I copied the pattern from an existing test (most probably `testRMWebAddress()` where it is actually required). But I will change it to Configuration and verify.

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (sanjaypujare) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: Good point. I will rewrite using `PowerMockito.mockStatic` (if possible) to prevent the global change

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (sanjaypujare) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java:200]: fixed

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (sanjaypujare) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: fixed

**Comment by sanjaypujare** [comment]
Both of @vrozov's comments addressed. The build failures are known and unrelated to this code. Jenkins failure is being addressed in https://github.com/apache/apex-core/pull/573.

**Comment by sanjaypujare** [comment]
Just looked up Apache site for voting/reviewing guidelines at http://www.apache.org/foundation/voting.html. The section on Voting has the following:

To prevent vetos from being used _capriciously_, they must be accompanied by a technical justification showing why the change is bad (opens a security exposure, negatively affects performance, etc. ). A veto without a justification is invalid and has no weight.

The thing to note is "... why the change is bad..." which excludes factors external to the change (such as build issues).

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (vrozov) [inline engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: Can it be moved to `@Before` or `@BeforeClass`?

Comment (sanjaypujare) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: No. Not all tests need that. But I will refactor this into a function that will be called from all the affected functions.

Comment (sanjaypujare) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: @vrozov Let me know your comments on the latest changes and the current state of this PR. The PR shows "vrozov requested changes" and you had a -1 based on CI issues (not related to this PR).

**Comment by vrozov** [comment]
@sanjaypujare It is not only me that have -1 on the PR - Jenkins and Travis have -1 on the PR as well. This PR as well as few other currently open PRs need to be rebased after #575 and #573 are merged. Once it is done and Jenkins/Travis approve PR, I'll dismiss my -1.

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (vrozov) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: Even though not all test use UserGroupInformation, will it affect them? Note that you mock UserGroupInformation for the entire unit test, not only for the tests that need it.

**Comment by sanjaypujare** [comment]
I don't agree with your reasoning but I can't do anything about it.

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (sanjaypujare) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: Yes, I am using PrepareForTest and RunWith annotations at class level because they could not be made to work at method level. There is more than one way, but I didn't think it was worth the time for me to refactor to address the `@Before` suggestion esp. at this stage.

**Comment by vrozov** [comment]
You may talk to one of committers to ask them to merge two PRs I mentioned.

File: engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java
Comment (vrozov) [reply engine/src/test/java/com/datatorrent/stram/client/StramClientUtilsTest.java]: The current way gives wrong impression that UserGroupInformation won't be modified unless `setupLoginTestUser()` is invoked. Also, should not it be static and `@BeforeClass`?

**Comment by vrozov** [comment]
@sanjaypujare rebase

**Comment by tweise** [comment]
@sanjaypujare rebase the PR

**Comment by sanjaypujare** [comment]
@tweise rebased

**Review by vrozov — CHANGES_REQUESTED** [review]
@sanjaypujare The build continues to fail, so keeping -1. Also, please address my comments regarding static initialization in the unit test.

**Comment by sanjaypujare** [comment]
@vrozov  I'll get back to this PR when I have more time. I'll close it for now.

**Comment by tweise** [comment]
To trigger CI build.

**Comment by tweise** [comment]
Last two Travis builds fail with
```
Tests in error: 
  StramClientUtilsTest.testEvalConfiguration:101 Â» NoClassDefFound Could not ini...
  StramClientUtilsTest.testEvalExpression:68 Â» NoClassDefFound Could not initial...
  StramClientUtilsTest.testVariableSubstitution:87 Â» ExceptionInInitializer
```
Previous Travis builds (different PRs) do not fail on that test:
https://travis-ci.org/apache/apex-core/pull_requests

**Comment by vrozov** [comment]
@sanjaypujare Please reopen when ready to work on the PR

**Pull request closed without merging**
